flink-core/src/main/java/org/apache/flink/api/common/typeutils/TypeSerializer.java

```diff
@@ -161,7 +162,93 @@
+	 * configuration if they guarantee forwards compatibility. For example, implementations that use serialization
+	 * frameworks with built-in serialization compatibility, such as <a href=https://thrift.apache.org/>Thrift</a> or
+	 * <a href=https://developers.google.com/protocol-buffers/>Protobuf</a>, is suitable for this usage pattern. By
+	 * returning the {@link ForwardCompatibleSerializationFormatConfig#INSTANCE}, this informs Flink that when managed
+	 * state serialized using this serializer is restored, there is no need to check for migration with the new
+	 * serializer for the same state. In other words, new serializers are always assumed to be fully compatible for the
+	 * serialized state.
+	 *
+	 * @see TypeSerializerConfigSnapshot
+	 * @see ForwardCompatibleSerializationFormatConfig
+	 *
+	 * @return snapshot of the serializer's current configuration.
+	 */
+	public abstract TypeSerializerConfigSnapshot snapshotConfiguration();
+
+	/**
+	 * Get the migration strategy to use this serializer based on the configuration snapshot of a preceding
+	 * serializer that was registered for serialization of the same managed state (if any - this method is only
+	 * relevant if this serializer is registered for serialization of managed state).
+	 *
+	 * <p>Implementations need to return the resolved migration strategy. The strategy can be one of the following:
+	 * <ul>
+	 *     <li>{@link MigrationStrategy#noMigration()}: this signals Flink that this serializer is compatible, or
+	 *     has been reconfigured to be compatible, to continue reading old data, and that the
+	 *     serialization schema remains the same. No migration needs to be performed.</li>
+	 *
+	 *     <li>{@link MigrationStrategy#migrateWithFallbackDeserializer(TypeSerializer)}: this signals Flink that
+	 *     migration needs to be performed, because this serializer is not compatible, or cannot be reconfigured to be
+	 *     compatible, for old data. Furthermore, in the case that the preceding serializer cannot be found or
+	 *     restored to read the old data, the provided fallback deserializer can be used.</li>
+	 *
+	 *     <li>{@link MigrationStrategy#migrate()}: this signals Flink that migration needs to be performed, because
+	 *     this serializer is not compatible, or cannot be reconfigured to be compatible, for old data.</li>
+	 * </ul>
+	 *
+	 * <p>This method is guaranteed to only be invoked if the preceding serializer's configuration snapshot is not the
+	 * singleton {@link ForwardCompatibleSerializationFormatConfig#INSTANCE} configuration. In such cases, Flink always
+	 * assume that the migration strategy is {@link MigrationStrategy#migrate()}.
+	 *
+	 * @see MigrationStrategy
+	 *
+	 * @param configSnapshot configuration snapshot of a preceding serializer for the same managed state
+	 *
+	 * @return the result of the reconfiguration.
+	 */
+	protected abstract MigrationStrategy<T> getMigrationStrategy(TypeSerializerConfigSnapshot configSnapshot);
```

Comment:
This method is currently doing (up to) 3 things: (maybe) reconfigure the serializer, give advise about if we need to convert, and (maybe) provide a backwards compatibe serializer. Maybe this can be told as a high level overview in the beginning of the doc, then all the details.

From this, I think also a name like `ensureCompatibility` returning a `CompatibilityResult` could be a better fit?